v2/comments_test.go

```diff
@@ -309,106 +319,143 @@ func TestParseTagKeyWithTagNames(t *testing.T) {
+		{input: "name(lower)", expectKey: "name", expectArgs: mkss("lower")},
+		{input: "name(CAPITAL)", expectKey: "name", expectArgs: mkss("CAPITAL")},
+		{input: "name(MiXeD)", expectKey: "name", expectArgs: mkss("MiXeD")},
+		{input: "name(mIxEd)", expectKey: "name", expectArgs: mkss("mIxEd")},
+		{input: "name(_under)", expectKey: "name", expectArgs: mkss("_under")},
+		{input: `name("hasQuotes")`, expectKey: "name", expectArgs: mkss("\"hasQuotes\"")},
+		{input: "name(`hasRawQuotes`)", expectKey: "name", expectArgs: mkss("`hasRawQuotes`")},
+		{input: "name(has space)", expectKey: "name", err: true},
+		{input: "name(has-dash)", expectKey: "name", err: true},
+		{input: "name(multiple, args)", expectKey: "name", err: true},
+		{input: "name(noClosingParen", expectKey: "name", err: true},
 	}
 	for _, tc := range cases {
-		ret, err := parseTagArgs(tc.input)
+		s := initTagKeyScanner(tc.input)
+		key, args, err := s.parseTagKey([]string{"name"})
+
 		if err != nil && tc.err == false {
 			t.Errorf("[%q]: expected success, got: %v", tc.input, err)
 			continue
 		}
 		if err == nil {
 			if tc.err == true {
-				t.Errorf("[%q]: expected failure, got: %q", tc.input, ret)
+				t.Errorf("[%q]: expected failure, got: %q", tc.input, key)
 				continue
 			}
-			if len(ret) != len(tc.expect) {
-				t.Errorf("[%q]: expected %d results, got: %q", tc.input, len(tc.expect), ret)
+			if key != tc.expectKey {
+				t.Errorf("[%q]\nexpected key: %q, got: %q", tc.input, tc.expectKey, key)
+			}
+			if len(args) != len(tc.expectArgs) {
+				t.Errorf("[%q]: expected %d args, got: %q", tc.input, len(tc.expectArgs), args)
 				continue
 			}
-			for i := range tc.expect {
-				if want, got := tc.expect[i], ret[i]; got != want {
+			for i := range tc.expectArgs {
+				if want, got := tc.expectArgs[i], args[i]; got != want {
 					t.Errorf("[%q]\nexpected %q, got %q", tc.input, want, got)
 				}
 			}
 		}
 	}
 }
 
-func TestSplitKeyValScanner(t *testing.T) {
+func TestParseJSON(t *testing.T) {
 	cases := []struct {
 		input string
-		key   string
-		val   string
+		err   bool
 	}{
-		{`foo=bar`, "foo", "bar"},
-		{`foo   =   bar`, "foo", "   bar"},
-		{`keyWithRaw(` + "`a=b`" + `)=value`, "keyWithRaw(`a=b`)", "value"},
-		{`noValue`, "noValue", ""},
-		{`rawKey=` + "`x=y`", "rawKey", "`x=y`"},
+		{
+			input: `[]`,
+		},
+		{
+			input: `{}`,
+		},
+		{
+			input: `[1]`,
+		},
+		{
+			input: `{"a":1}`,
+		},
+		{
+			input: `[1, 2]`,
+		},
+		{
+			input: `{"a": 1, "b": 2}`,
+		},
+		{
+			input: `1.1`,
+		},
+		{
+			input: `-4`,
+		},
+		{
+			input: `true`,
+		},
+		{
+			input: `false`,
+		},
+		{
+			input: `"string"`,
+		},
+		{
+			input: "null",
+		},
+
+		// invalid
+		{
+			input: `[1,]`,
+			err:   true,
+		},
+		{
+			input: `[1,]`,
+			err:   true,
+		},
+		{
+			input: `{"a":1,}`,
+			err:   true,
+		},
+		{
+			input: `{"a"`,
+			err:   true,
+		},
+		{
+			input: `"a`,
+			err:   true,
+		},
+		{
+			input: `UNKNOWN`,
+			err:   true,
+		},
```

Comment:
I am seeing some potential small issues with the parser around whitespace.  Using the below test cases:

		{
			// name:  "whitespace before }",
			input: `{"key":"value" }`,
		},
		{
			// name:  "whitespace before ]",
			input: `[1 ]`,
		},
		{
			// name:  "whitespace before comma in array",
			input: `[1 ,2]`,
		},
I get errors for these cases when I believe they are valid JSON.  Might make sense to document this in the supported JSON subset or get these tests passing.  I am fine with either option.

Reply:
Fixed